src/FffGcodeWriter.cpp

```diff
@@ -1660,57 +1732,93 @@ bool FffGcodeWriter::addSupportInfillToGCode(const SliceDataStorage& storage, La
-            support_infill_overlap = infill_extr_here.getSettingInMicrons("infill_overlap_mm"); // support lines area should be expanded outward to overlap with the boundary polygon
         }
 
-        int extra_infill_shift = 0;
-        bool use_endpieces = true;
-        Polygons* perimeter_gaps = nullptr;
-        double fill_angle = 0;
-        Infill infill_comp(support_pattern, island, offset_from_outline, support_line_width, support_line_distance, support_infill_overlap, fill_angle, z, extra_infill_shift, perimeter_gaps, infill_extr.getSettingBoolean("support_connect_zigzags"), use_endpieces);
-        Polygons support_polygons;
-        Polygons support_lines;
-        infill_comp.generate(support_polygons, support_lines);
-        if (support_lines.size() > 0 || support_polygons.size() > 0)
+        // process sub-areas in this support infill area with different densities
+        for (uint32_t density_idx = 0; density_idx < support_infill_part.gradual_infill_areas.size(); ++density_idx)
         {
-            setExtruder_addPrime(storage, gcode_layer, layer_nr, infill_extruder_nr_here); // only switch extruder if we're sure we're going to switch
-            gcode_layer.setIsInside(false); // going to print stuff outside print object, i.e. support
-            gcode_layer.addPolygonsByOptimizer(support_polygons, &gcode_layer.configs_storage.support_infill_config);
-            gcode_layer.addLinesByOptimizer(support_lines, &gcode_layer.configs_storage.support_infill_config, (support_pattern == EFillMethod::ZIG_ZAG)? SpaceFillType::PolyLines : SpaceFillType::Lines);
-            added = true;
+            if (support_infill_part.gradual_infill_areas[density_idx].empty())
+            {
+                continue;
+            }
+            const Polygons& support = support_infill_part.gradual_infill_areas[density_idx][0];
+
+            unsigned int density_factor = 2 << density_idx; // == pow(2, density_idx + 1)
+            int support_line_distance_here = support_line_distance * density_factor; // the highest density infill combines with the next to create a grid with density_factor 1
+            int support_shift = support_line_distance_here / 2;
+            if (density_idx == support_infill_part.gradual_infill_areas.size() - 1)
+            {
+                support_line_distance_here /= 2;
+            }
+
+            std::vector<PolygonsPart> support_islands = support.splitIntoParts();
+
+            for (unsigned int n = 0; n < support_islands.size(); ++n)
+            {
+                PolygonsPart& island = support_islands[n];
+
+                int offset_from_outline = 0;
+                int support_infill_overlap = 0; // support infill should not be expanded outward
+                if (support_pattern == EFillMethod::GRID || support_pattern == EFillMethod::TRIANGLES || support_pattern == EFillMethod::CONCENTRIC)
+                {
+                    offset_from_outline = -support_line_width;
```

Comment:
You shouldn't be performing this offset at this stage, but at the PolygonRenerator phase.

At this point you perform the offset to each density area separately, while you should be performing the offset before you generated the first plain infill area.
Doing it this way means you alter where the boundary between different density areas lies.

Moreover, now you first split the infill into density areas and thickness areas, meaning you have to apply the offset to a lot more areas.
If you apply the offset before you compute the density and thickness areas you only have to apply the offset once.

I suggest that `AreaSupport::generateGradualSupport` doesn't work with `supportAreas` (which are effectively the outlines) but with a new member variable `SupportInfillPart::infill_area` which is computed to be an offset from the last inset.

You could add a function `SupportAreas::generateInfillArea` which is called right after `generateOutlineInsets`.
Alternatively you could do the computation inside the same function as `generateOutlineInsets` and rename that function to something like `generateWallsAndInfill`.